policies/releasestrat.html

```diff
@@ -55,65 +56,20 @@ <h5>
 	  fixes. Before that, bug and security fixes will be applied
 	  as appropriate.</p>
 
-	  <p>The next version of OpenSSL will be 1.1.1 which will be an LTS release.
-	  This is currently in development and has a primary focus of implementing
-	  TLSv1.3. The RFC for TLSv1.3 has not yet been published by the IETF.
-	  OpenSSL 1.1.1 will not have its final release until that has happened.</p>
-
-	  <p>The draft release timetable for 1.1.1 is as follows. This may be
-          amended at any time as the need arises.</p>
-
+	  <p>The following stability rules apply to all changes made to code
+	  targetted for a version after 1.1.1:</p>
 	  <ul>
-	    <li>13th February 2018, alpha release 1 (pre1)</li>
-	    <li>27th February 2018, alpha release 2 (pre2)</li>
-	    <li>20th March 2018, beta release 1 (pre3)
-              <ul>
-	        <li>OpenSSL_1_1_1-stable created (feature freeze)</li>
-	        <li>master becomes basis for 1.1.2 or 1.2.0 (TBD)</li>
-	      </ul>
-	    <li>3rd April 2018, beta release 2 (pre4)</li>
-	    <li>17th April 2018, beta release 3 (pre5)</li>
-	    <li>1st May 2018, beta release 4 (pre6)</li>
-	    <li>29th May 2018, beta release 5 (pre7)</li>
-	    <li>19th June 2018, beta release 6 (pre8)</li>
-	    <li>Release readiness check following pre8 release (new release
-	        cycles added if required)</li>
+	    <li>No existing public interface can be removed or modified until its
+	    replacement has been in place in an LTS stable release and the original
+	    function has been documented as deprecated for at least 5 years.</li>
```

Comment:
So, for an API to be *modified*, it would have to be documented as deprecated? Not sure that makes sense. The most recent example of a possible API change is to `SSL_set_options()`, where it currently takes and returns a `long`. Changing that to a `uint64_t` allows for a consistent number (64) of options, where as right now, it depends on the compiler defintion of `long`. Obviously, changing the type breaks the ABI, but there's a subtle difference between changing an input parameter to 64bits, and the return value to 64bits. The first could be considered API (source) compatible, since the compiler will automatically extend the input to match function signature. But that doesn't happen for the return value.

Reply:
That's a good point. I was thinking of modifications more along the lines of adding or removing parameters. I'm not sure about changing types.

If you changed SSL_set_options() to take `uint64_t` (which is probably API compatible) - wouldn't that necessarily imply also changing SSL_get_options() to return a `uint64_t` which would be API breaking.
I'd be interested in @vdukhovni's thoughts.

Reply:
I suppose there is also the example of "const"ifying parameters - which we've generally allowed.

Reply:
The discussion here is about the API, not ABI, and the example of widening an integer parameter does not introduce an API break. Existing calls that pass a long continue to work when the target type changes to uint64_t, through the magic of C prototypes and integral promotion rules. Similarly, constification of non-pointer input parameters is harmless.

The goal here is to avoid abrupt API transitions where we have one API in OpenSSL version X and a different incompatible API in version X+1, with no common way to perform some functions across the two versions. The notion of X and X+1 is specified as a time range, since our downstream consumers don't necessarily adopt each and every version, but rather do batch upgrades at major plaform releases, ...  We can debate whether 5 years is "just right", or some other number is better, but the idea should be to provide plenty of opportunity for users doing upgrades to have a real migration path where their software compiles against both the old and the new, and then get to gradually fix the deprecation warnings before the next major upgrade.  That way an OS distribution can rebuild all its packages, and work on modernization of the API without having upgrading OpenSSL blocked on porting all the applications.

Reply:
I agree with Viktor -- we definitely need to provide the compat APIs in the previous version to allow for a smooth transition. In particular, I think the following text about adding opaquification accessors should also allow adding compat APIs.
(It doesn't sound like we need to debate the actual timeout right now, but I would personally go for more like 2-3 years than 5.)

Reply:
I clarified "modified" to say "or modified (in an API incompatible manner)"